stackslib/src/net/stackerdb/config.rs

```diff
@@ -123,40 +130,87 @@ impl StackerDBConfig {
-                    }
-                    if !func_return_type
-                        .admits_type(epoch, &returns)
-                        .unwrap_or(false)
-                    {
-                        let reason = format!("Contract function '{}' has an invalid return type: expected {:?}, got {:?}", name, func_return_type, returns);
-                        return Err(reason);
-                    }
-                }
-                _ => {
-                    let reason = format!("Contract function '{}' is not a fixed function", name);
-                    return Err(reason);
+            let FunctionType::Fixed(func) = func else {
+                return Err(format!("Function '{name}' must be a fixed function"));
+            };
+
+            if func.args.len() != expected_args.len() {
+                let reason = format!(
+                    "Function '{name}' has an invalid signature: it must have {} args",
+                    expected_args.len()
+                );
+                return Err(reason);
+            }
+            for (actual_arg, expected_arg) in func.args.iter().zip(expected_args.iter()) {
+                if !actual_arg
+                    .signature
+                    .admits_type(epoch, expected_arg)
+                    .unwrap_or(false)
+                {
+                    return Err(format!("Function '{name}' has an invalid argument type: expected {expected_arg}, got {actual_arg}"));
                 }
             }
+
+            if !expected_return
+                .admits_type(epoch, &func.returns)
+                .unwrap_or(false)
+            {
+                return Err(format!("Function '{name}' has an invalid return type: expected {expected_return}, got {}", &func.returns));
+            }
         }
         Ok(())
     }
 
-    /// Evaluate the contract to get its signer slots
+    fn parse_slot_entry(
+        entry: ClarityValue,
+        contract_id: &QualifiedContractIdentifier,
+    ) -> Result<(StacksAddress, u32), String> {
+        let ClarityValue::Tuple(slot_data) = entry else {
+            let reason = format!(
+                "StackerDB fn `{contract_id}.{STACKERDB_SLOTS_FUNCTION}` returned non-tuple slot entry",
+            );
+            return Err(reason);
+        };
+
+        let Ok(ClarityValue::Principal(signer_principal)) = slot_data.get("signer") else {
+            let reason = format!(
+                "StackerDB fn `{contract_id}.{STACKERDB_SLOTS_FUNCTION}` returned tuple without `signer` entry of type `principal`",
+            );
+            return Err(reason);
+        };
+
+        let Ok(ClarityValue::UInt(num_slots)) = slot_data.get("num-slots") else {
+            let reason = format!(
+                "StackerDB fn `{contract_id}.{STACKERDB_SLOTS_FUNCTION}` returned tuple without `num-slots` entry of type `uint`",
+            );
+            return Err(reason);
+        };
+
+        let num_slots = u32::try_from(*num_slots)
```

Comment:
I saw lots of earlier ""BUG: Failed to construct `{ signer: principal, num-slots: u64 }` tuple"," Is the num-slots meant to be a u32 or a u64? We do not seem to be consistent.

Reply:
It has to be a `u32` because the stackerdb system uses a `u32` to represent the total number of slots